### PR TITLE
[Feat]: Streaming Hidden-states Dataset

### PR DESCRIPTION
**Pull request by h-guo18**
### What does this PR do?

Type of change: new feature

**Design doc:** https://gist.github.com/h-guo18/241c94968b0591324c361d97cf995dd0
Jira ticket: https://jirasw.nvidia.com/browse/OMNIML-4341
Sandbox CI: https://gitlab-master.nvidia.com/omniml/integration/nmm-sandbox/-/jobs/327911555#L2228

Streaming hidden-states dataset: per-sample activations pulled from a live `vllm serve` over HTTP, replacing on-disk activation dumps.

Two axes for future extensions:
- **Backend** (`_fetch`): vLLM now; TRT-LLM / SGLang next.
- **Algorithm** (`_format`): Eagle now; distillation / probing next.
- **Sandbox CI**: https://gitlab-master.nvidia.com/omniml/integration/nmm-sandbox/-/merge_requests/169

Shared plumbing — async producer, token-level truncation to `training_seq_len`, loss-mask alignment, DDP via Accelerate's dispatcher (rank 0 fetches, broadcasts), circuit breaker, resume — lives in `StreamingDataset`. First instance: **`EagleVllmStreamingDataset`**.

API: `data.mode ∈ {online, offline, streaming}`; legacy configs auto-promote.

### Usage

```yaml
data:
  mode: streaming
  data_path: input_conversations/train.jsonl
  streaming_server_url: http://localhost:8000
  streaming_model_name: meta-llama/Llama-3.1-8B-Instruct
training:
  training_seq_len: 4096   # also caps the prompt sent to vllm
```

Requires `vllm serve` with `ExampleHiddenStatesConnector` and `dataloader_num_workers=0`.

End-to-end Slurm pipeline: `tools/launcher/examples/Qwen/Qwen3-8B/hf_streaming_eagle3.yaml`.

### Testing

- **Unit**: full-corpus invariant, rank-0-only iter, resume, circuit breaker, mocked-httpx integration.
- **E2E**: `launch_train.sh` against a stdlib `HTTPServer` mimicking the connector.
- **Smoke** (Qwen3-8B / 8×H100 / 4096 ultrachat samples, single epoch): train_loss 32 → 18, MT-Bench AR 1.003 → 1.20.

### TODO before un-drafting

- [ ] Observability counters (filtered, fetch failures, queue depth, latency).
- [ ] Changelog entry.
- [x] Add test in sandbox.

**Non-goals (v1):** multi-epoch streaming, cross-rank dynamic load balancing.

### Before your PR is "*Ready for review*"

- Backward compatible: ✅ (legacy configs auto-promote)
- New PIP dep: N/A
- New tests: ✅
- Changelog: ❌ (TODO above)
- Claude approval: ❌

## Summary by CodeRabbit

* **New Features**
  * Streaming training mode with server-backed hidden-state fetching, deterministic seed control, resume support, and streaming-specific dataset options (server, model, prefetch, shared storage).

* **Behavior / Bug Fixes**
  * Stronger mode validation; offline behavior derived from data mode; resume handling adjusted to avoid double-skip during streaming runs.

* **Tests**
  * End-to-end CI streaming test and expanded unit tests covering streaming, resume, DDP, determinism, and failure cases.

* **Infrastructure**
  * Launcher script and pipeline config for end-to-end streaming training.

[Review Change Stack](https://app.coderabbit.ai/change-stack/NVIDIA/Model-Optimizer/pull/1509?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by shengliangxu** [comment]
@h-guo18 I am working on configuration system/general dataset system, targeting all modelopt pipelines. We should discuss and work together to build one dataset infra.

File: modelopt/torch/speculative/plugins/hf_streaming_dataset.py [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+
+        ``safe_open(..., framework="pt").get_tensor`` materializes an independent
+        torch Tensor (not a view into the mmap'd file), so it is safe to unlink
+        right after the ``with`` block exits.
+        """
+        with safe_open(path, framework="pt") as f:
+            token_ids = f.get_tensor("token_ids")
+            hidden_states = f.get_tensor("hidden_states")  # [seq, n_layers, hidden]
+        with contextlib.suppress(OSError):
+            os.unlink(path)
```
[the remark this answers is not in the document]
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/hf_streaming_dataset.py:537]: addressed

File: tools/launcher/examples/Qwen/Qwen3-8B/hf_streaming_eagle3.yaml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      - training.num_train_epochs=1
+    environment:
+      - HF_MODEL_CKPT: <<global_vars.hf_model>>
+      # No spaces: nemo_run emits `export FOO=value` without quotes, so a
+      # space-separated value would be split by the shell.
+      - EAGLE_CAPTURE_IDS: "[2,18,33,36]"
+      - SERVE_GPU: "0"
+      - SERVE_TP: "1"
+      - SERVE_GPU_MEM_UTIL: "0.4"
+    slurm_config:
```
[the remark this answers is not in the document]
Comment (h-guo18) [reply tools/launcher/examples/Qwen/Qwen3-8B/hf_streaming_eagle3.yaml:64]: not relevant. guidline not accurate

File: modelopt/torch/speculative/plugins/hf_streaming_dataset.py [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+        labels[..., :-1] = input_ids[..., 1:]
+
+        return {
+            "input_ids": input_ids,
+            "base_model_hidden_states": base_model_hidden_states,
+            "aux_hidden_states": aux_hidden_states,
+            "attention_mask": torch.ones_like(input_ids),
+            "loss_mask": loss_mask,
+            "labels": labels,
+        }
```
[the remark this answers is not in the document]
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/hf_streaming_dataset.py:575]: Shape mismatch (wrong count) will surface as a tensor-shape error in the downstream spec-model forward — loud, just one layer later than ideal. The bigger concern raised here is wrong *order* (e.g. `[36,2,18,33]`), but order is fundamentally undetectable without the server returning the layer-id list it actually captured. In practice users derive `EAGLE_CAPTURE_IDS` from `default_eagle_aux_layer_ids(L) + [L]`, which is monotonic by construction, so this is unlikely. Leaving as-is; revisiting only if/when vLLM exposes the captured ids in its response.

File: modelopt/torch/speculative/plugins/hf_streaming_dataset.py [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+        return token_ids, hidden_states
+
+    @staticmethod
+    def _align_loss_mask(loss_mask: torch.Tensor, n: int) -> torch.Tensor:
+        """Trim or right-pad ``loss_mask`` to length ``n`` to match what vLLM tokenized to."""
+        if loss_mask.shape[0] > n:
+            return loss_mask[:n]
+        if loss_mask.shape[0] < n:
+            pad = torch.ones(n - loss_mask.shape[0], dtype=loss_mask.dtype)
+            return torch.cat([loss_mask, pad], dim=0)
```
[the remark this answers is not in the document]
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/hf_streaming_dataset.py]: Addressed: added a strict-equality check in `_fetch` enforcing that the server token sequence is the client tokenization verbatim, plus at most one decode-step token at the tail. Any drift (BOS prepend, chat templating, tokenizer mismatch) now raises `RuntimeError` instead of silently misaligning `loss_mask`. The `_align_loss_mask` tail-only behavior is correct under this enforced contract, and the misleading "BOS handling" comment has been rewritten to state the contract explicitly. Token-aware realignment (recomputing the mask against server tokens) is out of scope — the trainer assumes one tokenizer source of truth.

File: modelopt/torch/speculative/plugins/hf_streaming_dataset.py [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+
+    @staticmethod
+    def _align_loss_mask(loss_mask: torch.Tensor, n: int) -> torch.Tensor:
+        """Trim or right-pad ``loss_mask`` to length ``n`` (pad with 0; no client-side label)."""
+        if loss_mask.shape[0] > n:
+            return loss_mask[:n]
+        if loss_mask.shape[0] < n:
+            pad = torch.zeros(n - loss_mask.shape[0], dtype=loss_mask.dtype)
+            return torch.cat([loss_mask, pad], dim=0)
+        return loss_mask
```
[the remark this answers is not in the document]
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/hf_streaming_dataset.py:554]: Addressed: added a strict-equality check in `_fetch` enforcing that the server token sequence is the client tokenization verbatim, plus at most one decode-step token at the tail. Any drift (BOS prepend, chat templating, tokenizer mismatch) now raises `RuntimeError` instead of silently misaligning `loss_mask`. The `_align_loss_mask` tail-only behavior is correct under this enforced contract, and the misleading "BOS handling" comment has been rewritten to state the contract explicitly. Token-aware realignment (recomputing the mask against server tokens) is out of scope — the trainer assumes one tokenizer source of truth.

File: modelopt/torch/speculative/plugins/hf_streaming_dataset.py [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+from pathlib import Path
+from typing import TypedDict
+
+import httpx
+import torch
+from pydantic import BaseModel, ConfigDict, Field, field_validator
+from safetensors import safe_open
+from torch.utils.data import IterableDataset, get_worker_info
+from transformers import TrainerCallback
+from transformers.trainer_pt_utils import LabelSmoother
```
[the remark this answers is not in the document]
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/hf_streaming_dataset.py:53]: Partial-install CI is already green via two existing guards: `tests/unit/.../test_hf_streaming_dataset.py:34` does `pytest.importorskip("transformers")`, and the production import sites in `examples/speculative_decoding/{main.py,eagle_utils.py}` are lazy (`if mode == "streaming":` branch only). `plugins/__init__.py` does not import this module, so `import modelopt.torch.speculative.plugins` stays clean too. Adding this module to the `with import_plugin("transformers"):` block in `__init__.py` would be cosmetic / stylistic — leaving it out for now.

File: modelopt/torch/speculative/plugins/hf_streaming_dataset.py [hunk 7]
```diff
@@ -0,0 +1,558 @@
...
+        self._resume_skip = 0
+
+        indices = list(range(len(entries)))
+        random.Random(self.config.seed).shuffle(indices)
+        self.entries = [entries[i] for i in indices]
+        rank, world = dist_utils.rank(), dist_utils.size()
+        print(
+            f"[{type(self).__name__}] rank {rank}/{world}: "
+            f"holds {len(self.entries)} entries (full corpus; rank 0 fetches)"
+        )
```
[the remark this answers is not in the document]
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/hf_streaming_dataset.py]: Addressed: all 5 bare `print(...)` sites in this module now use `print_rank_0` / `warn_rank_0` from `modelopt.torch.utils`. The startup-log path (`__init__`) and the resume-callback log use `print_rank_0`; the three runtime warning paths (`_process` fetch exception, `_fetch` missing `hidden_states_path`, `_fetch` path-outside-root) use `warn_rank_0`. The redundant `if dist_utils.is_master():` guard in `StreamingResumeCallback` was dropped since `print_rank_0` already wraps it. Test fixture was updated to make the `rank` / `size` mocks accept `group=...` (needed because `is_master()` now reaches them via that kwarg).

File: modelopt/torch/speculative/plugins/hf_streaming_dataset.py [hunk 8]
```diff
@@ -0,0 +1,558 @@
...
+                        if outcome is True:
+                            consecutive_skips = 0
+                        elif outcome is False:
+                            consecutive_skips += 1
+                        # None -> entry unfit pre-fetch; server not at fault
+                    if consecutive_skips >= threshold:
+                        raise RuntimeError(
+                            f"{consecutive_skips} consecutive _fetch failures "
+                            f"in {type(self).__name__}; server likely down."
+                        )
```
[the remark this answers is not in the document]
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/hf_streaming_dataset.py:313]: Added a docstring note on `_produce` clarifying that the breaker has batch-level granularity (not per-task): when `asyncio.wait(..., FIRST_COMPLETED)` returns several tasks in the same loop turn, iteration over `done` (a set) is implementation-defined, so the final `consecutive_skips` reflects set order, not strict temporal order. For the breaker's actual purpose (detect a dead server), batch granularity is sufficient — not switching to `as_completed`-style or attaching timestamps.

File: modelopt/torch/speculative/plugins/hf_training_args.py [hunk 9]
```diff
@@ -63,6 +71,36 @@ def _check_sample_size(cls, v: int) -> int:
...
+            self.streaming_server_url
+            and self.streaming_model_name
+            and self.streaming_shared_storage_path
+        ):
+            raise ValueError(
+                "data.mode='streaming' requires data.streaming_server_url, "
+                "data.streaming_model_name, and data.streaming_shared_storage_path "
+                "(the trainer-side allowlist for paths returned by the vLLM server)"
+            )
+        return self
```
[the remark this answers is not in the document]
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/hf_training_args.py:96]: Addressed with the sentinel-default approach (option 2). `mode` is now `Literal[...] | None = None`; `_check_mode_requirements` auto-promotes only when `mode is None`. An explicit `mode: online` (or `offline` / `streaming`) is always respected — stale `streaming_server_url` / `offline_data_path` leftovers can no longer silently flip the resolved mode. Old YAMLs that omit `mode` keep the previous "infer from fields" behavior. Validator still resolves to one of the three string literals before returning, so downstream consumers (`recipe.eagle.eagle_offline = self.data.mode != "online"` etc.) are untouched. Verified across 7 scenarios incl. the bot's exact reproducer (explicit `online` + stale `streaming_server_url` → stays online).

**Comment by h-guo18** [comment]
> @h-guo18 I am working on configuration system/general dataset system, targeting all modelopt pipelines. We should discuss and work together to build one dataset infra.

Hi @shengliangxu , since this is a planned feature for 0.45 release, could we merge a version first and refactor later if necessary? Let me know if anything can be adjusted in this PR to make it easier later. Thanks!

File: modelopt/torch/speculative/plugins/hf_streaming_dataset.py [hunk 10]
```diff
@@ -0,0 +1,585 @@
...
+
+    def __init__(
+        self,
+        entries: list[dict],
+        tokenizer,
+        config: StreamingConfig | None = None,
+    ):
+        """Hold the *full* corpus on every rank; tokenize and fetch lazily during iteration.
+
+        DDP sharding is delegated to Accelerate, which wraps any ``IterableDataset`` in
```
Comment (AAnoosheh) [inline modelopt/torch/speculative/plugins/hf_streaming_dataset.py]: Can outputs not be streamed in a way such that each DDP rank can read just its own shard?

File: modelopt/torch/speculative/plugins/hf_streaming_dataset.py [hunk 11]
```diff
@@ -0,0 +1,585 @@
...
+        ``DataLoaderDispatcher`` (its default for iterables): rank 0 alone consumes the
+        underlying dataset and broadcasts each batch to the other ranks. Pre-sharding
+        the corpus here would shrink rank 0's view to 1/N and cap ``max_steps`` at
+        ``corpus / N**2``. We keep the full corpus on every rank and rely on
+        :meth:`__iter__`'s rank-0 guard so non-zero ranks don't burn server requests.
+
+        Shuffling with ``config.seed`` still happens (only rank 0's order matters, but
+        applying it everywhere keeps the contract simple and the seed reproducible).
+        """
```
Comment (AAnoosheh) [inline modelopt/torch/speculative/plugins/hf_streaming_dataset.py:164]: Can you add arguments in docstring? I don't know what entries is for example

File: modelopt/torch/speculative/plugins/hf_streaming_dataset.py (hunk 10 above)
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/hf_streaming_dataset.py]: Hi @AAnoosheh , the current logics is for HF accelerate, where rank0's data is dispatched by default (and there seems no concise way to turn it off). If we want to make it general for frameworks other than accelerate, I think we can move this data sharding logic to child class. What frameworks do we plan to support for distillation?

File: modelopt/torch/speculative/plugins/hf_streaming_dataset.py (hunk 11 above)
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/hf_streaming_dataset.py:164]: Added

**Review by ChenhanYu — APPROVED** [review]

File: modelopt/torch/speculative/plugins/hf_streaming_dataset.py (hunk 10 above)
Comment (AAnoosheh) [reply modelopt/torch/speculative/plugins/hf_streaming_dataset.py]: Ah I see. No worries then

**Review by AAnoosheh — APPROVED** [review]

**Review by shengliangxu — APPROVED** [review]
LGTM

File: tools/launcher/core.py [hunk 12]
```diff
@@ -253,13 +253,22 @@ def build_slurm_executor(
...
-        job_dir=job_dir,
-        identity=identity,
-    )
+    # When launching from a login node inside the cluster (host is localhost),
+    # use a LocalTunnel: nemo_run then runs sbatch and copies artifacts via local
+    # subprocess/shutil instead of ssh+rsync. This avoids flaky/hanging ssh-to-
+    # localhost (e.g. MaxStartups throttling on a shared login node, or clusters
+    # like HSG that are only reachable through an sss proxy so paramiko can't
+    # tunnel in from outside). For real remote hosts, keep the SSHTunnel.
+    if slurm_config.host in ("localhost", "127.0.0.1"):
```
Comment (h-guo18) [inline tools/launcher/core.py:262]: Hi @ChenhanYu , please take a look if this look good. Reason for change: some cluster is not available for remove-submission by launcher (e.g. HSG). Instead, we need to submit from HSG login node to compute node.
